Suboptimal support for "unittest.TestCase.subTest": missing exception/traceback information

```
from unittest import TestCase

class Tests(TestCase):
  def test_subTest(self):
    with self.subTest("subtest 1"):
      self.assertEqual(0, 1)
    with self.subTest("subtest 2"):
      self.assertEqual(0, 1)

  def test_topTest(self):
    self.assertEqual(0, 1)
```
produces the following `zope.testrunner` test result:
```
xmlsec3: PYTHONPATH=. zope-testrunner -v -s package --test-path=.
Running tests at level 1
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
  Running:
..

Failure in test test_topTest (package.tests.Tests)
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/dieter/tmp/xmlsec3/package/tests.py", line 11, in test_topTest
    self.assertEqual(0, 1)
  File "/usr/lib/python3.6/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.6/unittest/case.py", line 822, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 1


  Ran 2 tests with 3 failures, 0 errors and 0 skipped in 0.002 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.

Tests with failures:
   test_subTest (package.tests.Tests) [subtest 1]
   test_subTest (package.tests.Tests) [subtest 2]
   test_topTest (package.tests.Tests)
```

As you can see, the exception and traceback information is shown for a "noraml" test but unavailable for a "subTest". This makes the `unittest.TestCase.subTest` feature virtually unusable with `zope.testrunner`.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -5,6 +5,9 @@
 6.4.1 (unreleased)
 ==================
 
+- ``unittest.TestCase.subTest`` support
+  (`#91 <https://github.com/zopefoundation/zope.testrunner/issues/91>`_).
+
 
 6.4 (2024-02-27)
 ================

--- a/src/zope/testrunner/runner.py
+++ b/src/zope/testrunner/runner.py
@@ -989,9 +989,9 @@ class TestResult(unittest.TestResult):
 
         if self.options.post_mortem:
             if self.options.resume_layer:
-                self.options.output.error_with_banner("Can't post-mortem debug"
-                                                      " when running a layer"
-                                                      " as a subprocess!")
+                self.options.output.error_with_banner(
+                    "Can't post-mortem debug when running a layer as"
+                    " a subprocess!")
             else:
                 zope.testrunner.debug.post_mortem(exc_info)
         elif self.options.stop_on_error:

--- a/src/zope/testrunner/runner.py
+++ b/src/zope/testrunner/runner.py
@@ -974,6 +974,29 @@ class TestResult(unittest.TestResult):
         unittest.TestResult.addSkip(self, test, reason)
         self.options.output.test_skipped(test, reason)
 
+    def addSubTest(self, test, subtest, exc_info):
+        if exc_info is None:
+            return
+        stdout, stderr = self._restoreStdStreams()
+        outp = self.options.output
+        report = (outp.test_failure
+                  if issubclass(exc_info[0], test.failureException)
+                  else outp.test_error)
+        report(subtest, time.time() - self._start_time,
+               exc_info, stdout=stdout, stderr=stderr)
+
+        unittest.TestResult.addSubTest(self, test, subtest, exc_info)
+
+        if self.options.post_mortem:
+            if self.options.resume_layer:
+                self.options.output.error_with_banner("Can't post-mortem debug"
+                                                      " when running a layer"
+                                                      " as a subprocess!")
+            else:
+                zope.testrunner.debug.post_mortem(exc_info)
+        elif self.options.stop_on_error:
+            self.stop()
+
     def addError(self, test, exc_info):
         stdout, stderr = self._restoreStdStreams()
         self.options.output.test_error(test, time.time() - self._start_time,

--- a/src/zope/testrunner/tests/test_doctest.py
+++ b/src/zope/testrunner/tests/test_doctest.py
@@ -227,6 +227,7 @@ def test_suite():
             'testrunner-eggsupport.rst',
             'testrunner-stops-when-stop-on-error.rst',
             'testrunner-new-threads.rst',
+            'testrunner-subtest.rst',
             setUp=setUp, tearDown=tearDown,
             optionflags=optionflags,
             checker=checker),

--- a/src/zope/testrunner/tests/testrunner-ex/subtest.py
+++ b/src/zope/testrunner/tests/testrunner-ex/subtest.py
@@ -1,0 +1,25 @@
+##############################################################################
+#
+# Copyright (c) 2024 Zope Foundation and Contributors.
+# All Rights Reserved.
+#
+# This software is subject to the provisions of the Zope Public License,
+# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
+# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
+# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
+# FOR A PARTICULAR PURPOSE.
+#
+##############################################################################
+import unittest
+
+
+class TestSomething(unittest.TestCase):
+
+    def test_subTest(self):
+        with self.subTest("fail 1"):
+            self.assertEqual(0, 1)
+        with self.subTest("success"):
+            self.assertEqual(0, 0)
+        with self.subTest("fail 2"):
+            self.assertEqual(0, 1)

--- a/src/zope/testrunner/tests/testrunner-subtest.rst
+++ b/src/zope/testrunner/tests/testrunner-subtest.rst
@@ -1,0 +1,45 @@
+=====================
+ ``subTest`` Support
+=====================
+
+The testrunner supports ``unittest.TestCase.subTest``.
+
+    >>> import os.path, sys
+    >>> directory_with_tests = os.path.join(this_directory, 'testrunner-ex')
+    >>> from zope import testrunner
+    >>> defaults = [
+    ...     '--path', directory_with_tests,
+    ...     '--tests-pattern', 'subtest',
+    ...     ]
+
+    >>> from zope import testrunner
+
+    >>> sys.argv = 'test -vv'.split()
+    >>> x = testrunner.run_internal(defaults) # doctest: +ELLIPSIS
+    Running tests at level 1
+    ...
+      Running:
+     test_subTest (subtest.TestSomething.test_subTest)
+    <BLANKLINE>
+    Failure in test test_subTest (subtest.TestSomething.test_subTest) [fail 1]
+    Traceback (most recent call last):
+     testrunner-ex/subtest.py", Line NNN, in test_subTest
+        self.assertEqual(0, 1)
+    AssertionError: 0 != 1
+    <BLANKLINE>
+    <BLANKLINE>
+    <BLANKLINE>
+    Failure in test test_subTest (subtest.TestSomething.test_subTest) [fail 2]
+    Traceback (most recent call last):
+     testrunner-ex/subtest.py", Line NNN, in test_subTest
+        self.assertEqual(0, 1)
+    AssertionError: 0 != 1
+    <BLANKLINE>
+    <BLANKLINE>
+      Ran 1 tests with 2 failures, 0 errors and 0 skipped in N.NNN seconds.
+    Tearing down left over layers:
+      Tear down zope.testrunner.layer.UnitTests in N.NNN seconds.
+    <BLANKLINE>
+    Tests with failures:
+       test_subTest (subtest.TestSomething.test_subTest) [fail 1]
+       test_subTest (subtest.TestSomething.test_subTest) [fail 2]

--- a/src/zope/testrunner/tests/testrunner-subtest.rst
+++ b/src/zope/testrunner/tests/testrunner-subtest.rst
@@ -41,5 +41,5 @@ The testrunner supports ``unittest.TestCase.subTest``.
       Tear down zope.testrunner.layer.UnitTests in N.NNN seconds.
     <BLANKLINE>
     Tests with failures:
-       test_subTest (subtest.TestSomething.test_subTest) [fail 1]
-       test_subTest (subtest.TestSomething.test_subTest) [fail 2]
+       test_subTest (subtest.TestSomething...) [fail 1]
+       test_subTest (subtest.TestSomething...) [fail 2]

--- a/src/zope/testrunner/tests/testrunner-subtest.rst
+++ b/src/zope/testrunner/tests/testrunner-subtest.rst
@@ -19,7 +19,7 @@ The testrunner supports ``unittest.TestCase.subTest``.
     Running tests at level 1
     ...
       Running:
-     test_subTest (subtest.TestSomething.test_subTest)
+     test_subTest (subtest.TestSomething...)
     <BLANKLINE>
     Failure in test test_subTest (subtest.TestSomething...) [fail 1]
     Traceback (most recent call last):

--- a/src/zope/testrunner/tests/testrunner-subtest.rst
+++ b/src/zope/testrunner/tests/testrunner-subtest.rst
@@ -21,7 +21,7 @@ The testrunner supports ``unittest.TestCase.subTest``.
       Running:
      test_subTest (subtest.TestSomething.test_subTest)
     <BLANKLINE>
-    Failure in test test_subTest (subtest.TestSomething.test_subTest) [fail 1]
+    Failure in test test_subTest (subtest.TestSomething...) [fail 1]
     Traceback (most recent call last):
      testrunner-ex/subtest.py", Line NNN, in test_subTest
         self.assertEqual(0, 1)
@@ -29,7 +29,7 @@ The testrunner supports ``unittest.TestCase.subTest``.
     <BLANKLINE>
     <BLANKLINE>
     <BLANKLINE>
-    Failure in test test_subTest (subtest.TestSomething.test_subTest) [fail 2]
+    Failure in test test_subTest (subtest.TestSomething...) [fail 2]
     Traceback (most recent call last):
      testrunner-ex/subtest.py", Line NNN, in test_subTest
         self.assertEqual(0, 1)

